File: projects/assistant-email/src/email_assistant/memory/thread_store.py
"""SQLite-backed thread memory.

Why SQLite?
-----------
* Zero-ops (single file), works everywhere Python does.
* Transactional writes → safe even if a run is killed mid-way.
* Trivial to migrate to Postgres/pgvector later (see brief.md Memory Architecture).

Schema is intentionally tiny; long-term semantic memory belongs in a future
component (see Phase 3+ in dev-order.md).
"""

from __future__ import annotations

import sqlite3
import threading
from collections.abc import Iterator
from contextlib import contextmanager
from dataclasses import dataclass
from datetime import UTC, datetime
from pathlib import Path

from email_assistant.domain.email import Priority

_SCHEMA = """
CREATE TABLE IF NOT EXISTS threads (
    thread_id       TEXT PRIMARY KEY,
    subject         TEXT NOT NULL,
    sender          TEXT NOT NULL,
    last_seen_utc   TEXT NOT NULL,
    last_priority   TEXT NOT NULL,
    notified_utc    TEXT
);
CREATE INDEX IF NOT EXISTS ix_threads_last_seen ON threads(last_seen_utc);
"""


@dataclass(frozen=True, slots=True)
class ThreadRecord:
    """One row in the ``threads`` table."""

    thread_id: str
    subject: str
    sender: str
    last_seen_utc: datetime
    last_priority: Priority
    notified_utc: datetime | None = None


class ThreadStore:
    """Persistent record of every thread the assistant has processed.

    Thread-safe under multi-threaded access via an ``RLock`` around the
    connection. For our single-process cron model this is overkill but cheap.
    """

    def __init__(self, path: Path) -> None:
        self._path = path
        path.parent.mkdir(parents=True, exist_ok=True)
        self._lock = threading.RLock()
        self._conn = sqlite3.connect(path, check_same_thread=False)
        self._conn.row_factory = sqlite3.Row
        with self._tx() as cur:
            cur.executescript(_SCHEMA)

    # ------------------------------------------------------------------ API

    def close(self) -> None:
        with self._lock:
            self._conn.close()

    def upsert(self, record: ThreadRecord) -> None:
        """Insert or update a thread. ``notified_utc`` is preserved if not set."""
        with self._tx() as cur:
            cur.execute(
                """
                INSERT INTO threads
                    (thread_id, subject, sender, last_seen_utc, last_priority, notified_utc)
                VALUES (?, ?, ?, ?, ?, ?)
                ON CONFLICT(thread_id) DO UPDATE SET
                    subject        = excluded.subject,
                    sender         = excluded.sender,
                    last_seen_utc  = excluded.last_seen_utc,
                    last_priority  = excluded.last_priority,
                    notified_utc   = COALESCE(excluded.notified_utc, threads.notified_utc)
                """,
                (
                    record.thread_id,
                    record.subject,
                    record.sender,
                    _iso(record.last_seen_utc),
                    record.last_priority.value,
                    _iso(record.notified_utc) if record.notified_utc else None,
                ),
            )

    def get(self, thread_id: str) -> ThreadRecord | None:
        with self._lock:
            row = self._conn.execute(
                "SELECT * FROM threads WHERE thread_id = ?", (thread_id,)
            ).fetchone()
        return _row_to_record(row) if row else None

    def known_ids(self) -> set[str]:
        """Return the full set of thread IDs the store has ever seen."""
        with self._lock:
            rows = self._conn.execute("SELECT thread_id FROM threads").fetchall()
        return {r["thread_id"] for r in rows}

    def mark_notified(self, thread_id: str, when: datetime | None = None) -> None:
        """Record that we've notified the user about this thread."""
        when = when or datetime.now(UTC)
        with self._tx() as cur:
            cur.execute(
                "UPDATE threads SET notified_utc = ? WHERE thread_id = ?",
                (_iso(when), thread_id),
            )

    # -------------------------------------------------------------- internals

    @contextmanager
    def _tx(self) -> Iterator[sqlite3.Cursor]:
        with self._lock, self._conn:
            cur = self._conn.cursor()
            try:
                yield cur
            finally:
                cur.close()


# ---------------------------------------------------------------- module utils


def _iso(dt: datetime) -> str:
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=UTC)
    return dt.astimezone(UTC).isoformat()


def _row_to_record(row: sqlite3.Row) -> ThreadRecord:
    return ThreadRecord(
        thread_id=row["thread_id"],
        subject=row["subject"],
        sender=row["sender"],
        last_seen_utc=datetime.fromisoformat(row["last_seen_utc"]),
        last_priority=Priority(row["last_priority"]),
        notified_utc=(
            datetime.fromisoformat(row["notified_utc"]) if row["notified_utc"] else None
        ),
    )

File: projects/assistant-email-go/memory/threadstore.go
// Package memory persists per-thread state in SQLite.
//
// Why SQLite: zero-ops (single file), transactional writes safe under
// mid-run kills, trivial to migrate to Postgres later. Uses the pure-Go
// modernc.org/sqlite driver so builds stay CGO-free.
//
// Schema and semantics are a 1:1 port of the Python ThreadStore — including
// the load-bearing detail that notified_utc is preserved across upserts, so
// alert de-duplication survives repeated runs.
package memory

import (
	"database/sql"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"sync"
	"time"

	_ "modernc.org/sqlite"

	"github.com/m1981/temporal-go-agent-sdk/projects/assistant-email-go/domain"
)

const schema = `
CREATE TABLE IF NOT EXISTS threads (
    thread_id       TEXT PRIMARY KEY,
    subject         TEXT NOT NULL,
    sender          TEXT NOT NULL,
    last_seen_utc   TEXT NOT NULL,
    last_priority   TEXT NOT NULL,
    notified_utc    TEXT
);
CREATE INDEX IF NOT EXISTS ix_threads_last_seen ON threads(last_seen_utc);
`

// ThreadRecord is one row in the threads table.
type ThreadRecord struct {
	ThreadID     string
	Subject      string
	Sender       string
	LastSeenUTC  time.Time
	LastPriority domain.Priority
	NotifiedUTC  *time.Time // nil = never notified
}

// ThreadStore is the persistent record of every thread the assistant has processed.
type ThreadStore struct {
	mu sync.Mutex
	db *sql.DB
}

// Open creates parent directories, opens (or creates) the database, and
// applies the schema.
func Open(path string) (*ThreadStore, error) {
	if dir := filepath.Dir(path); dir != "." {
		if err := os.MkdirAll(dir, 0o755); err != nil {
			return nil, fmt.Errorf("threadstore: create dir: %w", err)
		}
	}
	db, err := sql.Open("sqlite", path)
	if err != nil {
		return nil, fmt.Errorf("threadstore: open: %w", err)
	}
	// A single connection serializes writes; ample for the 2-hourly digest.
	db.SetMaxOpenConns(1)
	if _, err := db.Exec(schema); err != nil {
		db.Close()
		return nil, fmt.Errorf("threadstore: apply schema: %w", err)
	}
	return &ThreadStore{db: db}, nil
}

func (s *ThreadStore) Close() error {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.db.Close()
}

// Upsert inserts or updates a thread. notified_utc is preserved when the
// incoming record does not set it.
func (s *ThreadStore) Upsert(rec ThreadRecord) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	var notified any
	if rec.NotifiedUTC != nil {
		notified = iso(*rec.NotifiedUTC)
	}
	_, err := s.db.Exec(`
        INSERT INTO threads
            (thread_id, subject, sender, last_seen_utc, last_priority, notified_utc)
        VALUES (?, ?, ?, ?, ?, ?)
        ON CONFLICT(thread_id) DO UPDATE SET
            subject        = excluded.subject,
            sender         = excluded.sender,
            last_seen_utc  = excluded.last_seen_utc,
            last_priority  = excluded.last_priority,
            notified_utc   = COALESCE(excluded.notified_utc, threads.notified_utc)`,
		rec.ThreadID, rec.Subject, rec.Sender, iso(rec.LastSeenUTC), string(rec.LastPriority), notified,
	)
	if err != nil {
		return fmt.Errorf("threadstore: upsert %s: %w", rec.ThreadID, err)
	}
	return nil
}

// Get returns the record for a thread, or (nil, nil) when unknown.
func (s *ThreadStore) Get(threadID string) (*ThreadRecord, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	row := s.db.QueryRow(
		`SELECT thread_id, subject, sender, last_seen_utc, last_priority, notified_utc
         FROM threads WHERE thread_id = ?`, threadID)

	var rec ThreadRecord
	var lastSeen string
	var priority string
	var notified sql.NullString
	err := row.Scan(&rec.ThreadID, &rec.Subject, &rec.Sender, &lastSeen, &priority, &notified)
	if errors.Is(err, sql.ErrNoRows) {
		return nil, nil
	}
	if err != nil {
		return nil, fmt.Errorf("threadstore: get %s: %w", threadID, err)
	}
	if rec.LastSeenUTC, err = parseISO(lastSeen); err != nil {
		return nil, fmt.Errorf("threadstore: get %s: %w", threadID, err)
	}
	rec.LastPriority = domain.Priority(priority)
	if notified.Valid {
		t, err := parseISO(notified.String)
		if err != nil {
			return nil, fmt.Errorf("threadstore: get %s: %w", threadID, err)
		}
		rec.NotifiedUTC = &t
	}
	return &rec, nil
}

// KnownIDs returns the full set of thread IDs the store has ever seen.
func (s *ThreadStore) KnownIDs() (map[string]bool, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	rows, err := s.db.Query(`SELECT thread_id FROM threads`)
	if err != nil {
		return nil, fmt.Errorf("threadstore: known ids: %w", err)
	}
	defer rows.Close()
	ids := make(map[string]bool)
	for rows.Next() {
		var id string
		if err := rows.Scan(&id); err != nil {
			return nil, fmt.Errorf("threadstore: known ids: %w", err)
		}
		ids[id] = true
	}
	return ids, rows.Err()
}

// MarkNotified records that the user has been alerted about a thread.
// A zero when means "now".
func (s *ThreadStore) MarkNotified(threadID string, when time.Time) error {
	if when.IsZero() {
		when = time.Now().UTC()
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	_, err := s.db.Exec(`UPDATE threads SET notified_utc = ? WHERE thread_id = ?`, iso(when), threadID)
	if err != nil {
		return fmt.Errorf("threadstore: mark notified %s: %w", threadID, err)
	}
	return nil
}

func iso(t time.Time) string { return t.UTC().Format(time.RFC3339Nano) }

func parseISO(s string) (time.Time, error) { return time.Parse(time.RFC3339Nano, s) }
